packages/react-next/src/components/Dropdown/Dropdown.base.tsx

```diff
@@ -39,30 +39,155 @@ import { IProcessedStyleSet } from '../../Styling';
   hasFocus: boolean;
   calloutRenderEdge?: RectangleEdge;
 }
 
-@withResponsiveMode
-export class DropdownBase extends React.Component<IDropdownInternalProps, IDropdownState> implements IDropdown {
+const DEFAULT_PROPS: Partial<IDropdownProps> = {
+  options: [],
+};
+
+function useSelectedItemsState({
+  defaultSelectedKeys,
+  selectedKeys,
+  defaultSelectedKey,
+  selectedKey,
+  options,
+  multiSelect,
+}: IDropdownProps) {
+  const oldOptions = usePrevious(options);
+  const [selectedIndices, setSelectedIndices] = React.useState<number[]>([]);
+
+  React.useEffect(() => {
+    /** Get all selected indexes for multi-select mode */
+    const getSelectedIndexes = (): number[] => {
+      if (selectedKeyPropToUse === undefined) {
+        if (multiSelect) {
+          return getAllSelectedIndices();
+        }
+        const selectedIndex = getSelectedIndex(null);
+        return selectedIndex !== -1 ? [selectedIndex] : [];
+      } else if (!Array.isArray(selectedKeyPropToUse)) {
+        const selectedIndex = getSelectedIndex(selectedKeyPropToUse);
+        return selectedIndex !== -1 ? [selectedIndex] : [];
+      }
+
+      const returnValue: number[] = [];
+      for (const key of selectedKeyPropToUse) {
+        const selectedIndex = getSelectedIndex(key);
+        selectedIndex !== -1 && returnValue.push(selectedIndex);
+      }
+      return returnValue;
+    };
+
+    const getAllSelectedIndices = (): number[] => {
+      return options
+        .map((option: IDropdownOption, index: number) => (option.selected ? index : -1))
+        .filter(index => index !== -1);
+    };
+
+    const getSelectedIndex = (searchKey: string | number | null | undefined): number => {
+      return findIndex(options, option => {
+        // eslint-disable-next-line eqeqeq
+        if (searchKey != null) {
+          return option.key === searchKey;
+        } else {
+          // eslint-disable-next-line deprecation/deprecation
+          return !!option.selected || !!option.isSelected;
+        }
+      });
+    };
+
+    // In controlled component usage where selectedKey is provided, update the selectedIndex
+    // state if the key or options change.
+    let selectedKeyPropToUse: string | number | string[] | number[] | null | undefined;
+
+    // this does a shallow compare (assumes options are pure), for the purposes of determining whether
+    // defaultSelectedKey/defaultSelectedKeys are respected.
+    const didOptionsChange = options !== oldOptions;
+
+    if (multiSelect) {
+      if (didOptionsChange && defaultSelectedKeys !== undefined) {
+        selectedKeyPropToUse = defaultSelectedKeys;
+      } else {
+        selectedKeyPropToUse = selectedKeys;
+      }
+    } else {
+      if (didOptionsChange && defaultSelectedKey !== undefined) {
+        selectedKeyPropToUse = defaultSelectedKey;
+      } else {
+        selectedKeyPropToUse = selectedKey;
+      }
+    }
+
+    if (
+      (selectedKeyPropToUse !== undefined || !oldOptions) &&
```

Comment:
Just to clarify, the new `|| !oldOptions` is to handle the old constructor logic, right?

Reply:
Yes